pkg/networkplugin-syncer/handlers/ovn/ovn_submariner_infrastructure.go

```diff
@@ -0,0 +1,254 @@
+	}
+
+	subGatewaySwitch := nbdb.LogicalSwitch{
+		Name: submarinerUpstreamSwitch,
+	}
+
+	subGatewayToLocalNetLsp := nbdb.LogicalSwitchPort{
+		Name:      submarinerUpstreamLocalnetPort,
+		Type:      "localnet",
+		Addresses: []string{"unknown"},
+		Options: map[string]string{
+			"network_name": SubmarinerUpstreamLocalnet,
+		},
+	}
+
+	err = libovsdbops.CreateOrUpdateLogicalSwitchPortsAndSwitch(ovn.nbdb, &subGatewaySwitch, &subGatewayToLocalNetLsp)
+	if err != nil {
+		return errors.Wrap(err, "Failed to Create submariner logical gateway switch and associated ports")
+	}
+
+	subJoinSwitch := nbdb.LogicalSwitch{
+		Name: submarinerDownstreamSwitch,
+	}
+
+	subJoinToSubRouterLsp := nbdb.LogicalSwitchPort{
+		Name: submarinerDownstreamSwPort,
+		Type: "router",
+		Options: map[string]string{
+			"router-port": submarinerDownstreamRPort,
+		},
+		Addresses: []string{"router"},
+	}
+
+	subJointoOvnRouterLsp := nbdb.LogicalSwitchPort{
+		Name: ovnClusterSubmarinerSwPort,
+		Type: "router",
+		Options: map[string]string{
+			"router-port": ovnClusterSubmarinerRPort,
+		},
+		Addresses: []string{"router"},
+	}
+
+	err = libovsdbops.CreateOrUpdateLogicalSwitchPortsAndSwitch(ovn.nbdb, &subJoinSwitch,
+		[]*nbdb.LogicalSwitchPort{&subJoinToSubRouterLsp, &subJointoOvnRouterLsp}...)
+	if err != nil {
+		return errors.Wrap(err, "Failed to Create submariner logical join switch and associated ports")
+	}
+
+	// At this point, we are missing the ovn_cluster_router policies and the
+	// local-to-remote & remote-to-local routes in submariner_router, that
+	// depends on endpoint details that we will receive via events.
+	return nil
+}
+
+// setupOvnClusterRouterLRPs configures the ovn cluster router's logical router policies.
+func (ovn *SyncHandler) setupOvnClusterRouterLRPs() error {
+	remoteSubnets := ovn.remoteEndpointSubnetSet()
+
+	if remoteSubnets.Size() == 0 {
+		klog.V(log.DEBUG).Info("No Remote Subnets to Process")
+		return nil
+	}
+
+	klog.V(log.DEBUG).Infof("Reconciling these raw remote subnets %v to logical router policies", remoteSubnets.Elements())
+
+	err := ovn.reconcileSubOvnLogicalRouterPolicies(remoteSubnets)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// associateSubmarinerRouterToChassis locks the submariner_router to a specific node.
+func (ovn *SyncHandler) associateSubmarinerRouterToChassis(chassis *sbdb.Chassis) error {
+	subLogicalRouter := nbdb.LogicalRouter{
+		Name: submarinerLogicalRouter,
+	}
+
+	submarinerRouter, err := libovsdbops.GetLogicalRouter(ovn.nbdb, &subLogicalRouter)
+	if err != nil {
+		return errors.Wrap(err, "Failed to fetch the ovn submariner router")
+	}
+
+	if submarinerRouter.Options == nil {
+		submarinerRouter.Options = map[string]string{}
+	}
+
+	submarinerRouter.Options["chassis"] = chassis.Name
+
+	err = libovsdbops.CreateOrUpdateLogicalRouter(ovn.nbdb, submarinerRouter)
+	if err != nil {
+		return errors.Wrap(err, "failed to set chassis option on ovn submariner router")
+	}
+
+	return nil
```

Comment:
Wouldn't this result in us always returning an error? And in this case we need to ensure that the `libovsdbops` function call  is wrapped here so that if it's encountered the user can understand the context

Reply:
No, `errors.Wrap` et al conveniently check if `err` is nil.

Reply:
Ah Nice I didn't know that :)